### PR TITLE
Open source the internal Fabric AutoML fork

### PR DESCRIPTION
**Pull request by thinkall**
## Summary

Consolidate the Fabric AutoML implementation into the public FLAML repository
so GitHub is the sole maintained library source. Internal Fabric build and
publication infrastructure may remain and consume this same source.

The original source import is retained, with current public `main` merged in
and the remaining internal compatibility fixes ported selectively. The
unrelated internal Git history is not merged into the public branch.

## Public compatibility and Fabric defaults

Normal `flaml` installation still requires only NumPy. `flaml[automl]` remains
usable without MLflow, Spark, or Fabric services.

`flaml.fabric.is_fabric_runtime()` uses the existing Fabric notebook/context
markers, without starting Spark or contacting services. Set
`FLAML_FABRIC_RUNTIME` before importing FLAML:

- `auto`: detect the runtime (default).
- `true` / `1`: explicitly enable Fabric defaults.
- `false` / `0`: preserve public defaults, including inside Fabric.

Explicit AutoML/Tune options take precedence.

| Behavior | Public defaults | Fabric defaults |
| --- | --- | --- |
| Model history | Disabled | Enabled unless overridden |
| MLflow integration | Requires an active run or enabled autologging | Also supports visualization history |
| MLflow tags | `flaml.*` | `synapseml.flaml.*` |
| Tunable feature engineering | Off unless configured | Off unless configured |

`mlflow_logging=False` disables FLAML's integration in both AutoML and Tune.
AutoML restores global and flavor-specific autologging settings. Explicit
pipeline registration logs and registers the current fitted pipeline, rather
than assuming the best trial contains model artifacts.
Autologging `log_models=False` suppresses automatic model and pipeline
artifacts, without preventing explicitly requested pipeline registration.

See [Fabric runtime behavior](https://github.com/microsoft/FLAML/blob/lijiang/open-source-internal-merge/flaml/fabric/README.md) for details and examples.

## Reconciled implementation

- Keep the Fabric feature-engineering, visualization, Optuna fANOVA, telemetry,
  data-preparation, and forecasting additions from the original import.
- Preserve current upstream categorical mappings, sample-weight/group
  splitting, memory-budget forwarding, sklearn compatibility, statsmodels
  fixes, Windows ARM64 support, and public examples.
- Port LightGBM 4.3 / sklearn 1.8 compatibility into both normal and zero-shot
  estimators.
- Reset AutoFE correctly on refit, failed fit, and no-op fit. Fit preprocessing
  only on training rows, separately for each cross-validation fold.
- Preserve irregular-series forecasting with statsmodels 0.15 by supplying
  explicit prediction indexes.
- Port the slim Fabric notebook dependencies and Python-specific joblib bounds
  into `pyproject.toml`; retain public PEP 621 packaging and the current public
  version (`2.7.0`).
- Preserve the Spark/non-Spark CI split and current Action SHA pins. Add
  isolated installed-wheel smoke jobs for the core and AutoML extras.
- Build release artifacts using the PEP 517 frontend.

## Scope deliberately excluded

Internal Azure DevOps pipelines, conda recipes, low-code generation templates,
internal feeds, runbooks, ownership metadata, and internal documentation-site
changes are not brought into the public repository. Public website
documentation and legacy AutoGen tests are preserved from current public main.

Fabric demo notebooks, benchmark artifacts, and third-party notices from the
original PR remain; they account for much of the PR's size.

## Validation

**Hosted CI passed on `9e3a59ca`.** [Run 35321067026](https://github.com/microsoft/FLAML/actions/runs/35321067026)
completed successfully across all 14 build-matrix jobs and all 4 isolated
public-install jobs. Pre-commit, CodeQL, and the other applicable checks also
passed; this head reports 27 successful checks and one intentionally skipped
release check.

**September 16 review update:** the nine new review comments are addressed,
and newer public-main changes (including per-fold resampling and version
`2.7.0`) are merged with conflicts resolved. The combined local parallel suite
passes 324 tests with one optional imbalanced-learn skip. Focused regressions
cover AutoFE reconstruction/custom learners, telemetry privacy and event
counts, Spark coordinates, and the ensemble skip policy. Both
development-container definitions use the supported public Python 3.10 image.
The image manifest is verified; a Docker daemon was not available for a local
container build. The Spark ensemble regression executed and passed (not skipped)
in all three hosted Spark jobs under Python 3.11, 3.12, and 3.13.

**September 18 update:** newer public-main changes are integrated, including
the Statsmodels future-horizon fix, zero-valued anonymous tuning metrics, and
README updates. The forecasting conflict is resolved without dropping either
upstream horizon semantics or irregular-series support. Real ARIMA/SARIMAX
regressions cover delayed/gapped requests, generated versus user-provided
regressors, preserved prediction indexes and target names. The targeted
parallel suite passes 136 tests with seven optional forecasting-library skips.
The first refreshed hosted run passed 12 of 14 matrix jobs; Windows 3.12/3.13
exposed numerical cancellation in a SARIMAX regression comparing different
forecast batch lengths. The fixture now compares the same terminal horizon
without changing production code, tolerances, or platform coverage; its
27-test parallel validation passes. The fresh hosted run now passes all matrix
jobs, including the previously failing Windows 3.12 and 3.13 jobs.

- 137 focused runtime, import, MLflow, AutoFE isolation, LightGBM, package
  metadata, and searcher regression tests passed.
- Registration tests additionally cover logging-disabled AutoFE pipelines,
  missing signatures, and a real registered-model reload.
- A follow-up opt-out regression verifies that an active parent run and its
  children receive no automatic model artifacts under `log_models=False`,
  while explicit registration still works. The 109 runtime/MLflow tests pass.
- The first full CI matrix identified native-ensemble AutoFE lookup, async
  best-run tag ordering, and tests relying on implicit Fabric history. These
  are corrected with scoped opt-ins and deterministic regressions, not skips.
  The related parallel-test batch passes 124 tests; the exact ensemble,
  history/pickle, and warm-start regressions also pass locally. The Spark
  history paths now pass in hosted jobs under Python 3.11, 3.12, and 3.13.
- Selected forecasting suites passed 105 tests, with 7 skips for optional
  forecasting libraries not installed locally. All 10 existing splitting
  tests passed. Existing AutoFE and AutoML/MLflow integration suites were also
  exercised.
- Actual LightGBM 4.3 + sklearn 1.8 fits and predictions passed for both
  AutoML and zero-shot estimators.
- Core and `[automl]` wheels were installed into clean environments and
  exercised outside the source checkout, with MLflow and Spark absent.
- `pre-commit run --all-files` and Actionlint validate the source/workflows.
- Source and wheel artifacts build successfully. This Windows host has a
  registry-related failure in the build frontend's isolated bootstrap;
  local artifact builds used `--no-isolation` inside a dedicated virtualenv.
  GitHub's Ubuntu publication workflow uses the standard isolated PEP 517
  build.

All nine review threads remain addressed and resolved, with no new unresolved
comments. Hosted CI is complete and the PR is conflict-free. Normal PR approval
is still required before merging; the PR has not been merged. Live Fabric
rollout and internal delivery integration remain separate from public CI validation.

### CODE REVIEW

File: test/spark/_init_spark.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+def _spark_jars_packages() -> str:
+    """Maven coordinates for the JARs added to the Spark session."""
+    return (
+        "com.microsoft.azure:synapseml_2.12:1.0.14,"
+        "org.apache.hadoop:hadoop-azure:3.3.5,"
+        "com.microsoft.azure:azure-storage:8.6.6,"
+        f"org.mlflow:mlflow-spark_2.12:{mlflow.__version__}"
+        if Version(mlflow.__version__) >= Version("2.9.0")
+        else f"org.mlflow:mlflow-spark:{mlflow.__version__}"
+    )
```
[the remark this answers is not in the document]
Comment (thinkall) [reply test/spark/_init_spark.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. The conditional MLflow coordinate is computed separately, then appended to the common SynapseML/Hadoop/Azure Storage coordinates. JVM-free regression tests cover MLflow 2.8.2, 2.9.0 and 2.22.1 together with the Spark 3/4 Scala selections, and pass with the normal repository fixtures.

File: flaml/fabric/autofe.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+        return empty_search_space
+
+    if raw_config == "off":
+        return empty_search_space
+
+    for estimator_name, estimator_class in task.estimators.items():
+        if estimator_class == learner_class:
+            break
+
+    if estimator_name.endswith("_spark"):
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/fabric/autofe.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. AutoML now passes the registered learner name through SearchState into parse_autofe_config. Direct callers without a name use an exact built-in match or the generic feature space, never the final unrelated entry. Tests cover a misleading final Spark entry, custom Spark-name exclusion, and propagation from AutoML.add_learner.

File: test/fabric/test_telemetry.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    flaml.tune.run(tune_func1, config={"x": flaml.tune.uniform(0, 1)}, num_samples=3, metric="metric", mode="min")
+    flaml.tune.run(tune_func2, config={"x": flaml.tune.uniform(0, 1)}, num_samples=3, metric="metric", mode="max")
+
+    """
+    Below assertions worked on my local machine, but not on Azure pipeline.
+    """
+    # captured_text = caplog.text
+    # assert len(re.findall("log_telemetry: flaml.automl", captured_text)) == 0
+    # assert len(re.findall("log_telemetry: flaml.tune", captured_text)) == 1
```
[the remark this answers is not in the document]
Comment (thinkall) [reply test/fabric/test_telemetry.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. The Tune test now runs two evaluations with a mocked reporter/runtime and an isolated one-time flag, asserting one flaml-tune event in Fabric and zero outside Fabric. It no longer depends on environment-specific captured logs or commented assertions.

File: test/spark/test_ensemble.py [hunk 4]
```diff
@@ -25,6 +26,12 @@
 else:
     skip_my_learner = True
 
+
+if sys.version_info >= (3, 11):
+    skip_py311 = True
+else:
+    skip_py311 = False
```
[the remark this answers is not in the document]
Comment (thinkall) [reply test/spark/test_ensemble.py]: Implemented in 3c4ee30ab256133d7150f7961f41602ae91de7e7: removed the historical Python>=3.11 skip, matching current public main, while retaining genuine prerequisite checks and the existing regression behavior. Setup/skip-policy tests cover Python 3.10-3.13. I am leaving this thread open until the fresh hosted Spark matrix confirms the ensemble test executes successfully; a local Spark runtime was unavailable.

File: test/fabric/test_telemetry.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    X_train = np.array([[1, 2], [3, 4], [5, 6]])
+    y_train = np.array([1, 2, 3])
+    automl_2.fit(X_train, y_train, max_iter=4)
+
+    """
+    Below assertions worked on my local machine, but not on Azure pipeline.
+    """
+    # captured_text = caplog.text
+    # assert len(re.findall("log_telemetry: flaml.automl", captured_text)) == 1
+    # assert len(re.findall("log_telemetry: flaml.tune", captured_text)) == 0
```
[the remark this answers is not in the document]
Comment (thinkall) [reply test/fabric/test_telemetry.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. The AutoML telemetry test now mocks the Fabric reporter/runtime, resets the one-time flag, performs two small AutoML fits, and asserts exactly one event with the expected activity and attributes. A non-Fabric case asserts no reporter calls. No validation remains commented out.

File: flaml/tune/tune.py [hunk 6]
```diff
@@ -483,11 +505,20 @@ def easy_objective(config):
...
     old_use_ray = _use_ray
     old_verbose = _verbose
     old_running_trial = _running_trial
     old_training_iteration = _training_iteration
-
+    kusto_logger.info(
+        f"tune.run: search_alg={search_alg}, metric={metric}, mode={mode}, time_budget_s={time_budget_s}, "
+        f"num_samples={num_samples}, automl_info={automl_info}, "
+        f"force_cancel={force_cancel}, mlflow_exp_name={mlflow_exp_name}, extra_tag={extra_tag}, "
+        f"use_spark={use_spark}, verbose={verbose}\nconfig={config}"
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/tune/tune.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. The centralized Tune startup message now contains only a configuration-entry count and boolean flags. It no longer serializes configuration contents/keys, user tags, experiment or metric names, or search-algorithm objects. The regression supplies a sentinel private value through config, extra_tag and mlflow_exp_name and verifies none reaches the central logger.

File: flaml/automl/ml.py [hunk 7]
```diff
@@ -434,13 +445,20 @@ def train_estimator(
     for param, value in fe_params.items():
         config_dic.pop(param)
 
+    autofe = None
+    if Featurization is not None and fe_params and X_train is not None:
+        autofe = Featurization(params=fe_params, task=task)
+        X_train = autofe.fit_transform(X_train, y_train)
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/automl/ml.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. AutoFE is constructed whenever fe.* choices exist; only fitting requires training data. Because the no-data branch otherwise replaces the wrapper with the raw estimator, it now returns an unfitted sklearn Pipeline containing AutoFE and the estimator. Records without AutoFE retain the existing raw-estimator return type. A regression exercises get_estimator_from_log, subsequent fitting, transformed data, and prediction.

File: test/fabric/test_autofe.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+from sklearn import __version__ as sklearn_version
+from sklearn import datasets
+from sklearn.model_selection import train_test_split
+
+from flaml import AutoML
+from flaml.automl import Featurization
+
+warnings.filterwarnings("ignore")
+
+skip_autofe = pytest.mark.skipif(sklearn_version < "1.3.0", reason="AutoFe requires sklearn>=1.3.0")
```
[the remark this answers is not in the document]
Comment (thinkall) [reply test/fabric/test_autofe.py]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7 using packaging.version.Version for both operands. The actual skip expression was checked against 1.2.2, 1.3.0, 1.9.0, 1.10.0 and 2.0.0; valid future versions are no longer skipped lexicographically.

File: Dockerfile [hunk 9]
```diff
@@ -1,5 +1,5 @@
 # basic setup
-FROM mcr.microsoft.com/devcontainers/python:3.10
+FROM mcr.microsoft.com/devcontainers/python:3.8
```
[the remark this answers is not in the document]
Comment (thinkall) [reply Dockerfile]: Fixed in 3c4ee30ab256133d7150f7961f41602ae91de7e7. Both Dockerfile and .devcontainer/Dockerfile now use the supported Python 3.10 image from public main, matching requires-python>=3.10. The image manifest was verified; a full local container build was unavailable because the Docker daemon is not running.

File: test/spark/test_ensemble.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thinkall) [reply test/spark/test_ensemble.py]: Hosted validation is now complete in https://github.com/microsoft/FLAML/actions/runs/35060029264 at a2b4b4bd0d10890e04ad9844dc5cbd8cf21b3390. The raw logs show test/spark/test_ensemble.py passing (not skipped) in all three Spark jobs under Python 3.11, 3.12 and 3.13. All 14 build-matrix jobs and all 4 public-install jobs passed. Closing this thread.
